src/theory/strings/regexp_operation.cpp

```diff
@@ -658,6 +659,7 @@ void RegExpOpr::firstChars(Node r, std::set<unsigned> &pcset, SetNodes &pvset)
         unsigned b = r[1].getConst<String>().front();
         b = String::convertUnsignedIntToCode(b);
         Assert(a < b);
+        Assert(b<std::numeric_limits<unsigned>::max());
```

Comment:
Ditto

Reply:
This is performance intensive code, so I'd prefer it be an Assert.

Reply:
We should probably make sure that we never get into this situation then (e.g. in the type checker or by rewriting upper bounds in ranges to be smaller than `std::numeric_limits<unsigned>::max()`).

Reply:
To expand a bit on why I am concerned: `a` and `b` are (potentially) directly coming from user input, so we have to somehow make sure that the user cannot specify the full `unsigned` range here, even in production.

Reply:
What prevents this is that `a` and `b` should be in the range of `0...String::num_codes()`, where `num_codes()` is a fixed bound that is controlled by us (currently 256).

However, I think this raises another (several) issues. The issue is whether there is a way for the user to constructor a string using an arbitrary unsigned.

If the constructor `String(const std::vector<unsigned>& s)` is available to the user, then technically we could have `a` and `b` be an arbitrary input.  We don't have any assertions in this constructor.  However, this constructor is only available to the user via the API currently.  And actually, taking this one step further, we need to check whether string constants passed to the theory solver are in the range `0...TheoryStringsRewriter::getAlphabetCardinality()`.

I think we need additional assertions in two ways:
- For strings constructed by arbitrary `std::vector<unsigned>`, we ensure that the corresponding code point of each unsigned `0...String::num_codes()`.
- For string constants registered to TheoryStrings, we ensure all code points are in the range `0...TheoryStringsRewriter::getAlphabetCardinality()`.